packages/controllers/src/subject-metadata/SubjectMetadataController.ts

```diff
@@ -0,0 +1,261 @@
+  name: string;
+  // TODO:TS4.4 make optional
+  extensionId: string | null;
+  host: string | null;
+  iconUrl: string | null;
+};
+
+type SubjectMetadataToAdd = PermissionSubjectMetadata & {
+  name: string;
+  extensionId?: string | null;
+  host?: string | null;
+  iconUrl?: string | null;
+} & Record<string, Json>;
+
+export type SubjectMetadataControllerState = {
+  subjectMetadata: Record<SubjectOrigin, SubjectMetadata>;
+};
+
+const stateMetadata = {
+  subjectMetadata: { persist: true, anonymous: true },
+};
+
+const defaultState: SubjectMetadataControllerState = {
+  subjectMetadata: {},
+};
+
+export type GetSubjectMetadataState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => SubjectMetadataControllerState;
+};
+
+export type ClearSubjectMetadataState = {
+  type: `${typeof controllerName}:clearState`;
+  handler: () => void;
+};
+
+export type TrimSubjectMetadataState = {
+  type: `${typeof controllerName}:trimMetadataState`;
+  handler: () => void;
+};
+
+export type SubjectMetadataControllerActions =
+  | GetSubjectMetadataState
+  | ClearSubjectMetadataState
+  | TrimSubjectMetadataState;
+
+export type SubjectMetadataStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [SubjectMetadataControllerState, Patch[]];
+};
+
+export type SubjectMetadataControllerEvents = SubjectMetadataStateChange;
+
+type AllowedActions = HasPermissions;
+
+export type SubjectMetadataControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  SubjectMetadataControllerActions | AllowedActions,
+  SubjectMetadataControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+type SubjectMetadataControllerOptions = {
+  messenger: SubjectMetadataControllerMessenger;
+  subjectCacheLimit: number;
+  state?: Partial<SubjectMetadataControllerState>;
+};
+
+/**
+ * A controller for storing metadata associated with permission subjects. More
+ * or less, a cache.
+ */
+export class SubjectMetadataController extends BaseController<
+  typeof controllerName,
+  SubjectMetadataControllerState,
+  SubjectMetadataControllerMessenger
+> {
+  private _subjectCacheLimit: number;
+
+  get subjectCacheLimit(): number {
+    return this._subjectCacheLimit;
+  }
+
+  private subjectsEncounteredSinceStartup: Set<string>;
+
+  private subjectHasPermissions: GenericPermissionController['hasPermissions'];
+
+  constructor({
+    messenger,
+    subjectCacheLimit,
+    state = {},
+  }: SubjectMetadataControllerOptions) {
+    if (!Number.isInteger(subjectCacheLimit) || subjectCacheLimit < 1) {
+      throw new Error(
+        `subjectCacheLimit must be a positive integer. Received: "${subjectCacheLimit}"`,
+      );
+    }
+
+    const hasPermissions = (origin: string) => {
+      return messenger.call('PermissionController:hasPermissions', origin);
+    };
+
+    if (state.subjectMetadata) {
+      SubjectMetadataController.trimMetadataState(
+        state as SubjectMetadataControllerState,
```

Comment:
This cast doesn't seem correct because this state is `Partial<SubjectMetadataControllerState>`, not `SubjectMetadataControllerState`.

It's true that `subjectMetadata` is definitely truthy, and that's the only property that we're using, but if other properties are defined in the future this could lead to errors.

As an alternative, can we update `trimMetadataState` to only accept the `subjectMetadata` object? The types would be easier that way.

Reply:
Fixed in: 62a6af4